src/ccs/multifolding.rs

```diff
@@ -0,0 +1,390 @@
+use super::cccs::{CCCSInstance, CCCSShape};
+use super::lcccs::LCCCS;
+use super::util::{compute_sum_Mz, VirtualPolynomial};
+use super::{CCSShape, CCSWitness};
+use crate::ccs::util::compute_all_sum_Mz_evals;
+use crate::hypercube::BooleanHypercube;
+use crate::spartan::math::Math;
+use crate::spartan::polynomial::MultilinearPolynomial;
+use crate::utils::bit_decompose;
+use crate::{
+  constants::{BN_LIMB_WIDTH, BN_N_LIMBS, NUM_FE_FOR_RO, NUM_HASH_BITS},
+  errors::NovaError,
+  gadgets::{
+    nonnative::{bignat::nat_to_limbs, util::f_to_nat},
+    utils::scalar_as_base,
+  },
+  r1cs::{R1CSInstance, R1CSShape, R1CSWitness, R1CS},
+  traits::{
+    commitment::CommitmentEngineTrait, commitment::CommitmentTrait, AbsorbInROTrait, Group, ROTrait,
+  },
+  utils::*,
+  Commitment, CommitmentKey, CE,
+};
+use bitvec::vec;
+use core::{cmp::max, marker::PhantomData};
+use ff::{Field, PrimeField};
+use flate2::{write::ZlibEncoder, Compression};
+use itertools::concat;
+use rayon::prelude::*;
+use serde::{Deserialize, Serialize};
+use sha3::{Digest, Sha3_256};
+use std::ops::{Add, Mul};
+use std::sync::Arc;
+
+// XXX: THe idea is to have Multifolding as IVC instance in the future, holding the main CCS
+// instances. Then the rest of CCS, CCCS, LCCCS hold references to it.
+// Is our single source of data.
+#[derive(Debug)]
+pub struct Multifolding<G: Group> {
+  ccs: CCSShape<G>,
+  ccs_mle: Vec<MultilinearPolynomial<G::Scalar>>,
```

Comment:
If multilinear polynomials are represented over the Boolean hypercube, then there is no difference between the original representation and the MLE-encoded representation. Is it necessary to have two separate ways to storing the same data?

I wonder if we can define certain MLE-related traits and just implement them on `CCSShape` so there is no need for separately tracking MLE-encoded data.